server/applicationset/applicationset.go

```diff
@@ -148,6 +169,108 @@ func (s *Server) List(ctx context.Context, q *applicationset.ApplicationSetListQ
 
 }
 
+func (s *Server) Watch(q *applicationset.ApplicationSetWatchQuery, ws applicationset.ApplicationSetService_WatchServer) error {
+	ctx := ws.Context()
+	namespace := s.appsetNamespaceOrDefault(q.AppsetNamespace)
+
+	if !s.isNamespaceEnabled(namespace) {
+		return security.NamespaceNotPermittedError(namespace)
+	}
+
+	projects := map[string]bool{}
+	for _, project := range q.GetProjects() {
+		projects[project] = true
+	}
+
+	claims := ws.Context().Value("claims")
+	selector, err := labels.Parse(q.GetSelector())
+	if err != nil {
+		return fmt.Errorf("error parsing the selector: %w", err)
+	}
+
+	logCtx := log.NewEntry(log.New())
+	if q.Name != "" {
+		logCtx = logCtx.WithField("application", q.GetName())
+	}
+
+	minVersion := 0
+	if q.GetResourceVersion() != "" {
+		if minVersion, err = strconv.Atoi(q.GetResourceVersion()); err != nil {
+			minVersion = 0
+		}
+	}
+
+	// sendIfPermitted is a helper to send the applicationset to the client's streaming channel if the
+	// caller has RBAC privileges permissions to view it
+	sendIfPermitted := func(ctx context.Context, a v1alpha1.ApplicationSet, eventType watch.EventType) {
+		if !s.isApplicationSetPermitted(selector, minVersion, claims, q.GetName(), namespace, projects, a) {
+			return
+		}
+
+		err := ws.Send(&v1alpha1.ApplicationSetWatchEvent{
+			Type:           eventType,
+			ApplicationSet: a,
+		})
+		if err != nil {
+			logCtx.Warnf("Unable to send stream message: %v", err)
+			return
+		}
+	}
+
+	events := make(chan *v1alpha1.ApplicationSetWatchEvent, watchAPIBufferSize)
+	// Mimic watch API behavior: send ADDED events if no resource version provided
+	// If watch API is executed for one application when emit event even if resource version is provided
+	// This is required since single app watch API is used for during operations like app syncing and it is
+	// critical to never miss events.
+	if q.GetResourceVersion() == "" || q.GetName() != "" {
+		appsets, err := s.appsetLister.List(selector)
+		if err != nil {
+			return fmt.Errorf("error listing appssets with selector: %w", err)
+		}
+		sort.Slice(appsets, func(i, j int) bool {
+			return appsets[i].QualifiedName() < appsets[j].QualifiedName()
+		})
+		for i := range appsets {
+			sendIfPermitted(ctx, *appsets[i], watch.Added)
+		}
+	}
+	unsubscribe := s.appsetBroadcaster.Subscribe(events)
+	defer unsubscribe()
+	for {
+		select {
+		case event := <-events:
+			sendIfPermitted(ctx, event.ApplicationSet, event.Type)
+		case <-ws.Context().Done():
+			return nil
+		}
+	}
+}
+
+func (s *Server) isApplicationSetPermitted(selector labels.Selector, minVersion int, claims any, appsetName, appsetNs string, projects map[string]bool, a v1alpha1.ApplicationSet) bool {
+	if len(projects) > 0 && !projects[a.Spec.Template.Spec.GetProject()] {
+		return false
```

Comment:
Maybe we should add debug logs to all conditions of this function so Argo CD admins can understand the reason why the AppSet is not permitted.